tests/test-froidure-pin-matrix.cpp

```diff
@@ -28,266 +28,201 @@
-      REQUIRE(S.position(S.generator(0)) == 0);
-      REQUIRE(S.contains(S.generator(0)));
-
-      auto x = to_matrix<Mat>(sr, {{2, 2, 0}, {1, 0, 0}, {0, 0, 0}});
-      REQUIRE(S.position(x) == UNDEFINED);
-      REQUIRE(!S.contains(x));
-      x.product_inplace_no_checks(S.generator(0), S.generator(0));
-      REQUIRE(S.position(x) == 2);
-      REQUIRE(S.contains(x));
-    }
 
-    // NTPSemiring(11, 3);
-    template <typename Mat>
-    void test007(NTPSemiring<> const* sr = nullptr) {
-      auto             rg = ReportGuard(REPORT);
-      FroidurePin<Mat> S;
-      S.add_generator(to_matrix<Mat>(sr, {{2, 1, 0}, {10, 0, 0}, {1, 2, 1}}));
-      S.add_generator(to_matrix<Mat>(sr, {{10, 0, 0}, {0, 1, 0}, {1, 1, 0}}));
-
-      REQUIRE(S.size() == 86);
-      REQUIRE(S.degree() == 3);
-      REQUIRE(S.number_of_idempotents() == 10);
-      REQUIRE(S.number_of_generators() == 2);
-      REQUIRE(S.number_of_rules() == 16);
-
-      REQUIRE(S[0] == S.generator(0));
-      REQUIRE(S.position(S.generator(0)) == 0);
-      REQUIRE(S.contains(S.generator(0)));
-
-      auto x = to_matrix<Mat>(sr, {{2, 2, 0}, {1, 0, 0}, {0, 0, 0}});
-      REQUIRE(S.position(x) == UNDEFINED);
-      REQUIRE(!S.contains(x));
-      x.product_inplace_no_checks(S.generator(1), S.generator(0));
-      REQUIRE(S.position(x) == 4);
-      REQUIRE(S.contains(x));
-    }
-  }  // namespace
-
-  LIBSEMIGROUPS_TEST_CASE_V3("FroidurePin<MaxPlusMat<2>>",
-                             "040",
-                             "Example 000",
-                             "[quick][froidure-pin][matrix]") {
-    test000<MaxPlusMat<2>>();
+  LIBSEMIGROUPS_TEMPLATE_TEST_CASE("FroidurePin",
+                                   "040",
+                                   "Example 000",
+                                   "[quick][froidure-pin][matrix]",
+                                   MaxPlusMat<2>,
+                                   MaxPlusMat<>) {
+    auto                  rg = ReportGuard(REPORT);
+    FroidurePin<TestType> S;
+    S.add_generator(to_matrix<TestType>({{0, -4}, {-4, -1}}));
+    S.add_generator(to_matrix<TestType>({{0, -3}, {-3, -1}}));
+
+    REQUIRE(S.size() == 26);
+    REQUIRE(S.degree() == 2);
+    REQUIRE(S.number_of_idempotents() == 4);
+    REQUIRE(S.number_of_generators() == 2);
+    REQUIRE(S.number_of_rules() == 9);
+    REQUIRE(S[0] == S.generator(0));
+    REQUIRE(S[1] == S.generator(1));
+
+    REQUIRE(S.position(S.generator(0)) == 0);
+    REQUIRE(S.contains(S.generator(0)));
+
+    REQUIRE(S.position(S.generator(1)) == 1);
+    REQUIRE(S.contains(S.generator(1)));
+
+    TestType x({{-2, 2}, {-1, 0}});
+    REQUIRE(S.position(x) == UNDEFINED);
+    REQUIRE(!S.contains(x));
+    x.product_inplace_no_checks(S.generator(1), S.generator(1));
+    REQUIRE(S.position(x) == 5);
+    REQUIRE(S.contains(x));
+
+    x = TestType({{-2, 2, 0}, {-1, 0, 0}, {0, 0, 0}});
+    REQUIRE(S.position(x) == UNDEFINED);
+    REQUIRE(!S.contains(x));
```

Comment:
These are the lines that are causing asan to fail. I notice that, in the original `test000` function they were commented out.

Reply:
Yeah, here when `TestType` is a static 2x2 matrix, the constructor tries to access/assign values out of bounds, which I think causes the asan failure.